amt.go

```diff
@@ -53,41 +53,83 @@ func WrapBlockstore(bs blockstore.Blockstore) Blocks {
 	return &bstoreWrapper{bs}
 }
 
-func (bw *bstoreWrapper) Get(c cid.Cid, out cbg.CBORUnmarshaler) error {
-	b, err := bw.bs.Get(c)
+// Note: ctx not used.
+func (bw *bstoreWrapper) Get(ctx context.Context, c cid.Cid, out interface{}) error {
+	blk, err := bw.bs.Get(c)
 	if err != nil {
 		return err
 	}
 
-	if err := out.UnmarshalCBOR(bytes.NewReader(b.RawData())); err != nil {
-		return err
+	cu, ok := out.(cbg.CBORUnmarshaler)
+	if ok {
+		return cu.UnmarshalCBOR(bytes.NewReader(blk.RawData()))
 	}
+	return cbor.DecodeInto(blk.RawData(), out)
+}
 
-	return nil
+type cidProvider interface {
+	Cid() cid.Cid
 }
 
-func (bw *bstoreWrapper) Put(obj cbg.CBORMarshaler) (cid.Cid, error) {
-	buf := new(bytes.Buffer)
-	if err := obj.MarshalCBOR(buf); err != nil {
-		return cid.Undef, err
+// Note: ctx not used.
+func (bw *bstoreWrapper) Put(ctx context.Context, v interface{}) (cid.Cid, error) {
+	mhType := uint64(mh.BLAKE2B_MIN + 31)
```

Comment:
this logic is duplicated from the hamt code, right? Can we just import it from there instead?

Reply:
What if we extract these shared Get and Put methods to another package, and then have the HAMT and AMT import the new code?